### PR TITLE
feat(pages): implement cw20 pages

### LINKED ISSUE #100: Create CW20 contract page
image

image

### PR DESCRIPTION
**Pull request by grikomsn**
Fixes #100 💯

## Description

This PR implements CW20 contract pages for instantiate, query, and execute.

## Changes

List any technical changes.

- [x] create cw20 instantiate page
- [x] create cw20 query page
- [ ] create cw20 execute page
- [x] implement cw20 instantiate page form logics

## Screenshots

**Instantiate (WIP)**

image

**Query**

image

## Testing Steps

As a reviewer, what steps should I take to verify this is working correctly?

- create cw20 token from instantiate page
- test cw20 in query page
- initiate cw20 contract executions in execute page

## Links

\-

### CODE REVIEW

**Comment by findolor** [comment]
I think I messed up some things with the development branch...

**Comment by grikomsn** [comment]
@findolor, I have restored the issues after the branch deletion.

re: contract pages, execute page is still wip and create page has no form logics implemented. Will iterate on that this week.

File: components/LinkTabs.data.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    href: '/contracts/cw20/instantiate',
+  },
+  {
+    title: 'Query',
+    description: `Lorem ipsum dolor sit amet consectetur adipisicing elit.`,
+    href: '/contracts/cw20/query',
+  },
+  {
+    title: 'Execute',
+    description: `Lorem ipsum dolor sit amet consectetur adipisicing elit.`,
```
Comment (RyanWarner) [inline components/LinkTabs.data.ts:16]: @findolor I'd like to replace this lorem ipsum with real copy.

Could you help me get started? In your own words, can you describe each of these actions (instantiate, query, execute)?

I can help refine and edit.

File: pages/contracts/cw20/execute.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+      <h1 className="font-heading text-4xl font-bold">CW20 Tokens</h1>
+      <p>
+        CW20 is a specification for fungible tokens based on CosmWasm. Learn
+        more in the{' '}
+        <Anchor
+          href={links['Docs CW20']}
+          className="font-bold text-plumbus hover:underline"
+        >
+          documentation page
```
Comment (RyanWarner) [inline pages/contracts/cw20/execute.tsx:22]:
```suggestion
          documentation
```

File: pages/contracts/cw20/instantiate.tsx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        CW20 is a specification for fungible tokens based on CosmWasm. Learn
+        more in the{' '}
+        <Anchor
+          href={links['Docs CW20']}
+          className="font-bold text-plumbus hover:underline"
+        >
+          documentation page
+        </Anchor>
+        .
+      </p>
```
Comment (RyanWarner) [inline pages/contracts/cw20/instantiate.tsx:34]: @grikomsn This is duplicated code, could we abstract into a component?

File: pages/contracts/cw20/query.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        CW20 is a specification for fungible tokens based on CosmWasm. Learn
+        more in the{' '}
+        <Anchor
+          href={links['Docs CW20']}
+          className="font-bold text-plumbus hover:underline"
+        >
+          documentation page
+        </Anchor>
+        .
+      </p>
```
Comment (RyanWarner) [inline pages/contracts/cw20/query.tsx:72]: Another instance of this code that could be DRYd up.

File: components/LinkTabs.data.ts (hunk 1 above)
Comment (findolor) [reply components/LinkTabs.data.ts:16]: Ok will write some things and share it with you

Comment (grikomsn) [reply components/LinkTabs.data.ts]: Did some updates in f60e55edad5d6c62b45f1f0b21408fe29799dc10, but do update accordingly.

Comment (findolor) [reply components/LinkTabs.data.ts]: @RyanWarner 

Instantiate: Initialize and create new instances of cw20 contract.
Query: Query any cw20 contract address to get data
Execute: Invoke a method on any cw20 contract address

I think this is a good start, somethings along these lines should do

**Comment by findolor** [comment]
@grikomsn Everything works with instantiate now. There is just one thing to mention.

Initial balances is a required field in the object. Currently we only have our address in the array but we should be able to add more addresses if we want. Instead of using our address, let's change it to be an empty array since the user might not want to add their address to the list. 

So in the end it'll be required field with 2 inputs. Address and amount. Once we fill and click add, it'll push that object to the list. We also should be able to view which addresses and amounts we have in the list.

**Review by findolor — APPROVED** [review]
LGTM 🚀 

As I've said I'll put together a list of missing things. We could tackle them in another PR.

Thanks a lot for the quick development! @grikomsn  @RyanWarner

**Pull request merged**